File: 0187_repeated_dna_sequences/hash_set/cpp/solution.cpp
#include <algorithm>
#include <cassert>
#include <iostream>
#include <string>
#include <unordered_set>
#include <vector>

namespace {

constexpr int kWindowSize = 10;

std::vector<std::string> findRepeatedDnaSequences(const std::string& s) {
    if (s.size() < kWindowSize) {
        return {};
    }

    std::unordered_set<std::string> seen;
    std::unordered_set<std::string> repeated;
    std::vector<std::string> result;

    for (int start = 0; start <= static_cast<int>(s.size()) - kWindowSize; ++start) {
        std::string fragment = s.substr(start, kWindowSize);
        if (seen.contains(fragment)) {
            if (!repeated.contains(fragment)) {
                repeated.insert(fragment);
                result.push_back(fragment);
            }
        } else {
            seen.insert(fragment);
        }
    }

    return result;
}

void assertSameContents(std::vector<std::string> actual, std::vector<std::string> expected) {
    std::sort(actual.begin(), actual.end());
    std::sort(expected.begin(), expected.end());
    assert(actual == expected);
}

}  // namespace

int main() {
    assertSameContents(
        findRepeatedDnaSequences("AAAAACCCCCAAAAACCCCCCAAAAAGGGTTT"),
        {"AAAAACCCCC", "CCCCCAAAAA"});
    assertSameContents(findRepeatedDnaSequences("AAAAAAAAAAAAA"), {"AAAAAAAAAA"});
    assertSameContents(findRepeatedDnaSequences("ACGT"), {});
    assertSameContents(findRepeatedDnaSequences("ACGTACGTAC"), {});
    assertSameContents(findRepeatedDnaSequences("AAAAAAAAAAAA"), {"AAAAAAAAAA"});

    std::cout << "All tests passed!\n";
    return 0;
}

File: 0187_repeated_dna_sequences/hash_set/go/solution.go
package repeateddnasequences

const windowSize = 10

func findRepeatedDnaSequences(s string) []string {
	if len(s) < windowSize {
		return []string{}
	}

	seen := make(map[string]struct{}, len(s))
	repeated := make(map[string]struct{})
	result := make([]string, 0)

	for start := 0; start <= len(s)-windowSize; start++ {
		fragment := s[start : start+windowSize]
		if _, ok := seen[fragment]; ok {
			if _, alreadyAdded := repeated[fragment]; !alreadyAdded {
				repeated[fragment] = struct{}{}
				result = append(result, fragment)
			}
			continue
		}
		seen[fragment] = struct{}{}
	}

	return result
}
